pkg/addon/templateagent/values.go

```diff
@@ -154,3 +157,49 @@ func (a *CRDTemplateAgentAddon) sortValueKeys(value addonfactory.Values) []strin
 func hubKubeconfigPath() string {
 	return "/managed/hub-kubeconfig/kubeconfig"
 }
+
+const (
+	// ClusterImageRegistriesAnnotation represent the annotation key for image registries, the vale of the annotation
+	// should be a json string like this:
+	//
+	// {
+	//   "registries": [
+	//     {
+	//       "source": "quay.io/ocm",
+	//       "mirrors": "quay.io/open-cluster-management"
+	//     }
+	//   ]
+	// }
+	// TODO: move this to the api repo
+	ClusterImageRegistriesAnnotation = "open-cluster-management.io/image-registries"
+)
+
+func GetAddOnRegistriesPrivateValuesFromClusterAnnotation(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (addonfactory.Values, error) {
+	values := map[string]interface{}{}
+	annotations := cluster.GetAnnotations()
+	klog.V(4).Infof("Try to get image registries from annotation %v", annotations[ClusterImageRegistriesAnnotation])
+	if len(annotations[ClusterImageRegistriesAnnotation]) == 0 {
+		return values, nil
+	}
+	type ImageRegistries struct {
+		Registries []addonapiv1alpha1.ImageMirror `json:"registries"`
+	}
+
+	imageRegistries := ImageRegistries{}
+	err := json.Unmarshal([]byte(annotations[ClusterImageRegistriesAnnotation]), &imageRegistries)
```

Comment:
can it be unmarshalled correctly if there is pullSecret filed in the annotation?

Reply:
yes